### ISSUE TITLE
Allow a target to be always rebuilt

### ISSUE THREAD

**Issue by stoggi** [jpakkane/meson#30]
It would be handy to be able to mark a target as `build_always` which rebuilds it regardless of dependencies.

A typical use case is generating version header files from version control.

 * Should we use the PHONY target as a dependency?
 * How should we set Ninja's restat argument?

It would be nice if we could make it work always without users needing to specify extra arguments in their custom target definitions.

**Comment by axxel** [comment]
One could imagine a new special kind of 'target' explicitly for this purpose: some custom command that gets executed iff the used version control system says there has been a change. This would require to explicitly support different version control systems, which might very well make it a non-starter right there. But if not, it might be better implemented as a special configure step rather than a custom target, I mean something that reruns meson.py as if the meson.build file had changed but only if it makes sens.

I am not too convinced of this idea myself but it crossed my mind after having read the 'without users needing to specify extra arguments' comment and wondering how that might be possible.

**Comment by stoggi** [comment]
Hm, yes it would be nice to be able to specify your software version in your meson.build file, giving it a source, eg. hardcoded string, file, or external command. The calculated version string would then be available as an object in all your meson.buid scripts to be used however the user sees fit. In my example I would then simply pass it to the commithash.c file using the c_args.

Although, it would still be nice for a custom target to be able to always be rebuilt for other uses too.

**Comment by Xenopathic** [comment]
Once the frontend is in place, with some sort of parsing for a flag such as `build_always: true`, then the backend is certainly possible. The Ninja backend has a `PHONY` target that is 'always out-of-date', so anything depending on it is always rebuilt. Then all that is required is to hook up the frontend flag with the backend target generation. Have a look at 339d2a0a2b70b4d61f75e56be7d800079de3fb3a for how the backend is done

**Comment by jpakkane** [comment]
Ninja has the restat property for exactly this use case. The way it works is that your generator script would create the new source file, compare it against the old one and if they are identical, would not touch commithash.c at all. Ninja would notice that the time stamp has not updated and would infer that the file was up to date and would not recompile it (or any targets that depend on it). Similar functionality should be duplicatable with other backends, too (CMake does it, for example).

**Comment by axxel** [comment]
I've prepared a solution for this issue (took me a lot longer than anticipated, looks like I'm getting old and slow :-/), which I'll commit to a fork soon for everyone interested to have a look at. It is in part pretty much what @stoggi posted in his original e-mail two weeks ago: the option to allow a custom target to be rebuild every time ninja is run, by adding the `phony` dependency. I used the proposed `build_always` keyword. As @jpakkane noted, the `restat` property is helpful here but for custom targets this flag is already set, so nothing to do here. Unfortunately, this ninja feature does not spare you writing your own script that will not overwrite the output file in case nothing has changed.

I looked around like other people deal with this seemingly 'common' problem of including some version string based on their VCS (version control system) status into their build. I found a very nice blog post about this very issue: http://neugierig.org/software/blog/2014/11/binary-revisions.html. It has some interesting background on how the LLVM and the Chromium people deal with it.

I had the following feature requests for my implementation:
 1. The version string has to be up-to-date on *every* build.
 1. Therefore it has to be *fast*, especially during no-op builds (i.e. no changes on your VCS side).
 1. Don't require to write custom scripts to get something straight forward, like configuring a `version.c` file by replacing a `@VCS_INFO@` variable with the revision number of your mercurial source tree.
 1. Ideally support multiple VCS systems out of the box.
 1. Make the most common use case very easy but still allow pretty much every behaviour to be customized/overwritten.

To give an example on how the meson.build file looks like for simplest use case (see feature request 3.):
````
version_c = vcs_configure('version.c', 'version.c.in', vcs_info : vcs_info())
exe = executable('my_exe', 'main.c', version_c)
````
Given that your meson.build file is placed inside a WC (working copy) of a git or mercurial repository, this is all you need to have `version.c` always contain the current revision of your WC. If there is no change to your WC, then `ninja` returns `ninja: no work to do.` If there is a change to your WC state like an update but none of your source files changed, `version.c` will be regenerated.
If you also need to be up-to-date regarding some kind of dirty-flag in your WC (e.g. hg adds a '+' to the revision), you'll have to add the `build_always : true` flag to your `vcs_info()` statement. This then causes a small run-time overhead even in a no-op build (like 0.3s on my system).

To give an example of a full blown use case with all bells and whistles, imagine you want to have a custom version string from a mercurial repository looking like this: `<revision>-<branch>`. You write a script containing the line `hg id -nb | sed 's/ /-/'` and call it `get_vcs_info.sh`. Suppose you have a `version.c.in` like before but also have a dpkg `control.in` file for later packing up a binary debian package that contains the variable `@PACKAGE_VERSION@` which you'd like to replace. Here is the meson.build part:
````
my_vcs_info = vcs_info('my_vcs_info', vcs_cmd : 'get_vcs_info.sh', build_always : true)
version_c = vcs_configure('version.c', 'version.c.in',
    vcs_info : my_vcs_info
)
control_file = vcs_configure(
    input : 'debian/control.in',
    output : 'control',
    config_var_name : 'PACKAGE_VERSION',
    vcs_info : my_vcs_info
)
````
Both `vcs_info` and `vcs_configure` functions return opaque objects (CustomTargetHolder at the moment) and you can have multiple invocations of both.

I hope I sparked your interest and this work will end up in some form in meson, so someone will have some use from it. I'll report back as soon as I have set up my meson fork - hopefully later this evening.

**Comment by jpakkane** [comment]
Thanks for looking into this, however there are a few design issues here. Meson is very strict about adding new top level commands. The code you have listed adds two and they don't seem to provide much that would differ from a custom target. I'd be very hesitant to add those without really good reasons.

It might be better to solve this problem in a more general sense than just thinking about vcs files. One thing which I have been pondering is to create a wrapper binary that transparently checks the output of any command and preserves the output files' timestamps if the contents don't change. For an example of a wrapper binary see rustrunner.py.

If this script were done, then the generation command would be something like this:

    x = custom_target('foo', output : 'xxx.c', ... command : ['foogen.sh', 'input', 'xxx.c']

And the actual command that would be written into the Ninja file would be this (it would be phony and restat):

    statpreserver.py xxx.c foogen.sh input xxx.c

In this way people could write whatever scripts they wanted to without caring and it would still work.

**Comment by axxel** [comment]
First some comments I was about to post regarding my fork (then some comments regarding your feedback to the code below):
It is working as described earlier with the following limitations:
 * lots of error checking missing (I found the existing parameter checking to be quite repetitive and partly redundant so I wanted to see what can be done about that before making another copy of it. I think along the lines of the existing `validate_arguments` function.)
 * the `build_always` part is only implemented for the ninja backend
 * missing support for cascaded configuration: `configure_file` currently removes *all* `@XYZ@` strings which prevents a subsequent `vcs_configure` of a `configure_file` output
 * due to the limits of the `svn` command, I need to add some 'selector' facility to get to the interesting part of the `svn info` output
 * no test cases, yet (not quite clear how to test given the 'dependency' on different VCS systems)

Now to your feedback: You are right in two aspects
 * both commands do something that might more or less be achievable with a custom target (but this is also true for pretty much every other build target ;))
 * the proposed `statpreserver.py` is more general and fixes part of what I tried to achieve (basically the update_vcs_info function maybe minus proper cwd handling)

But that alone would not give you the following features of my proposal:
 * having a solution for the default case (use some configure-type operation to replace a simple variable) without needing to resort back to custom scripts, which I find problematic given the intended multi-platform support. Plus it will force everyone interested in this type of support to figure it out herself or find the proper wiki documentation and then copy'n'paste the script infrastructure from there. Why not have it ready to use in place?
 * truly no-op builds for the usual suspects of VCS systems
 * automatic detection of the used VCS and it's WC root to support the no-op builds

So what would definitely be still missing is some kind of `configure_target`, I mean something along the lines of the `configure_vcs_info` function that performs a build-time configuration step. I thought about this general solution but since the passing of the configure information (a string as the contents of a file) in this case seemed too specific, I went for the 'do something specific in an easy to understand and easy to type' way.

I did not anticipate that you'd be 'very strict about adding new top level commands'. Could you give me some hints as to why you think they could be harmful? What made the `jar` or the `pkgconfig_gen` target special enough to justify their addition over some custom_target + custom script as you suggested here?

Having that said, the proposed statpreserver.py is certainly a good idea from my point of view as it would solve part of what I tried to do here plus it would be usable elsewhere. But I did no understand how you would deal with the PHONY dependency. Having every custom_target depend on PHONY is certainly a bad idea, even if the rebuild-chain would stop there due to the statpreserver.py + restat. So you'll still need the user to decide and explicitly state whether something should be build_always or not. Are you already working on the statpreserver.py or should I give it a go, use it in my proposal and see where we are?

**Comment by axxel** [comment]
Appendix: I just realized that the statpreserver.py alone will not help me prevent the need for a custom script just to get the output of `git describe` into a file. It would either be necessary to do in fact allow shell commands like `git describe > @OUTPUT0@` or have some special flag that makes the statpreserver.py do this stdout redirection for me. I'd strongly prefer the former as it is a lot more concise and is what every programmer is familiar with anyway. Without either, the statpreserver.py would merely allow me to replace 3 lines of code in my proposal.

**Comment by jpakkane** [comment]
Originally the jar command did not exist but instead shared_library etc were used. However it turned out that jar is a special kind of output because several of the keyword arguments of shared_library are meaningless in int (soversion etc) and conversely it had its own unique keyword requirements (such as manifest). Because of these reasons it was made its own command.

If we step a bit further away from the vcs details and look at the bigger picture, we need a general way of generating sources with restat requirements and always_rebuild. Thus it seems like a good approach to first do the general solution. Then when we have used that for a while for vcs work are we in a better position to say whether Meson should have built in support for it. Note that you will most likely always need a wrapper script because, for example, different backends might invoke the command from different subdirectories.

As far as allowing shell syntax inside Meson goes, disallowing that has been a conscious design decision. The main reason for this is that on some platforms (most prominently Windows) a posixish shell is not available. Another issue is that if you permit this, people will start writing shell pipelines inside their build definitions. If you look at build definitions of popular projects, be they CMake or Autotools, this is exactly what happens and it makes those build definitions hard to understand, debug, modify and port to new platforms. Driving this functionality into external scripts is a bit more work up front, but makes maintainability massively simpler.

**Comment by axxel** [comment]
*about the shell syntax*: my proposal does not depend on it as the necessary wrapper script to perform the redirect is part of the implementation. It would only be necessary if I tried to replace that part with your suggested statpreserver.py idea. But also in general, if the only thing you need is a stdout redirect, then allowing that is a lot more understandable, debugable and portable than your suggested way of writing platform dependent helper scripts and distributing `if build.name() == 'windows'` type of statements across your build definition, which in my opinion are a real weak spot in the otherwise pretty clean and concise 'look' of a meson build file. I'd generally avoid the necessity of platform dependent code whenever possible. Only if you have an output that you generate only on a specific platform anyway, should you need to go that route, certainly not for things like this vcs-feature here.

*about the bigger picture*: since apparently this vcs type use case is the first that requires the restat+always_rebuild feature, what other use cases, requiring a more general solution, do you have in mind? As you can see in the code, the suggested solution is agnostic of the backend with regards to the path the command (here the meson_vcs_info.py) is invoked from, so you will *not* need a custom wrapper script to get the described feature set when using a different backend.

If I stick with feature request 1 and 3 while dropping 4 from the list and somewhat compromise on 2 and 5 but on the other hand remove the 'vcs' flavor, hence make it a general command, I could see a solution with (one) new top-level command `configure_target` that does more or less what my `vcs_configure` does but without the `vcs_info` part and an obligatory `var` parameter. It could look like this:
````
version_c = configure_target('version.c', 'version.c.in', var : 'VCS_INFO',
       command : ['git', 'describe'])
````
It would run the command in the current source directory (like commandrunner.py does), take the command's stdout and configure the input, i.e. replacing the `var` string with it. Alternatively to the command input, it could be passed a filename or a CustomTargetHolder to read the contents from. It should also implement the touch-output-only-if-different behaviour, as this is a trivial addition and would then not require yet another wrapper. It should still support the optional always_rebuild flag. Would you consider this as an acceptable extension to meson?

I spent already more time on this than is warranted with respect to the benefit I'd get by replacing my existing cmake build system. But I like your project and hoped I could contribute something useful for others, too. But it is of course your decision how you'd like to have it progress and I wish you success with it, because I really think you have a point about the necessity for a build system that does not suck. ;)

**Comment by jpakkane** [comment]
I'd like to apologize for not giving this feature the attention it deserves. Working on the Wrap portions took all my time in addition to some personal real life issues. Thank you for keeping this alive and submitting you comments, they have been very helpful.

To get this issue back on proper track I wrote a suggested implementation of the general rebuild solution. It is now in trunk along with a unit test. I used your first commit as a guideline (and marked you as a contributor). The properties of the solution are as follows:

- meson.build does not have any branching based on platform
- implementation is in Python so automatically multiplatform (Meson does the necessary magic to launch the script properly, even on Windows)
- statpreserver.py is not there, instead the script does it manually
- no change build time on my machine is 0.03s
- works both when the git directory is there and when it is not (that is, when compiling from a tarball as e.g. when building distro packages)

The last one is especially important because in order to support it seems that you either need branching in your meson.build or a wrapper script. 

This does not necessarily need to be the final solution, but having it there gives us a better understanding to determine if and how new things need to be added.

**Comment by jpakkane** [comment]
Since this bug was originally about always rebuilding and that is now landed, I'm going to close this issue. However I'm going to open a new one to discuss what to do with the automatic vcs support.

**Comment by rulatir** [comment]
VCS tagging is not the only use case for this. In general, dependency discovery is Hard But Sometimes Circumventable, and the capability requested here would provide general power to hook such circumvention mechanisms into the build process.

For example, determining which files a docker image build depends on is hard: it requires parsing the Dockerfile and making note of every ADD and COPY, and crossing your fingers and praying that it works. However docker builds that don't actually change the image are pretty fast thanks to docker's internal caching of layers. This opens an opportunity for circumventing the difficult dependency discovery and just "depending" on the hash of the entire source tree, which can be efficiently retrieved from a version control system and stored in a file. However we need the ability to have this source hash target always built; otherwise we would have to wrap the build system invocation in a script that first deletes the file, which is a sad thing to have to do.
